[AST] Eliminate IfConfigStmt

rintaro opened the pull request:
~~**NOTE: This conflicts with #7955 , I will rebase this PR when it's merged.**~~ landed

Resolves: [SR-4426](https://bugs.swift.org/browse/SR-4426)

`#if` in statements position doesn't have to be `Stmt`.
Modeling it as `IfConfigDecl` simplify the code and AST.

* Make `IfConfigDecl` be able to hold `ASTNode` elements
* Parse `#if` as `IfConfigDecl`
* Stop enclosing toplevel `#if` into `TopLevelCodeDecl`.
* ~New methods on `IfConfigDecl` for retrieving active elements as `Decl`s. This replaces `IfConfigDecl::getActiveMembers()`.~
* Eliminate `IfConfigStmt`

```swift
#if true
  #if true
  func foo() {}
  #endif
#else
  func bar() {}
#endif
```
is now dumped as:
```
(source_file
  (if_config_decl
    #if: active
      (boolean_literal_expr type='<null>' value=true)
      (elements
        (if_config_decl
          #if: active
            (boolean_literal_expr type='<null>' value=true)
            (elements
              (func_decl "foo()" ...)))
        (func_decl "foo()" ...))
    #else:
      (elements
        (func_decl "bar()" ...)))
  (func_decl "foo()" ...))
```

rintaro committed:
[Parse] Don't Hoist IfConfig{Decl,Stmt} out of IfConfig{Decl,Stmt}

rintaro committed:
[AST] Eliminate IfConfigStmt

Resolves: https://bugs.swift.org/browse/SR-4426

* Make IfConfigDecl be able to hold ASTNodes
* Parse #if as IfConfigDecl
* Stop enclosing toplevel #if into TopLevelCodeDecl.
* Eliminate IfConfigStmt

rintaro commented:
@slavapestov @jrose-apple 
What do you think?

rintaro commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

rintaro commented on include/swift/AST/Decl.h:
```diff
@@ -2026,31 +2027,64 @@ class SerializedTopLevelCodeDeclContext : public SerializedLocalDeclContext {
+  /// All elements must be Decl.
+  const AsDeclRange getAsDecls() const {
+    return AsDeclRange(getActiveClauseElements(), AsDeclWithSkippingIfConfig());
+  }
```
These new methods looks overkill since they only have one client.

rintaro commented on include/swift/AST/Stmt.h:
```diff
@@ -24,6 +24,7 @@
 #include "swift/AST/IfConfigClause.h"
 #include "swift/AST/TypeAlignments.h"
 #include "swift/Basic/NullablePtr.h"
+#include "swift/Basic/STLExtras.h"
```
Accidental addition.

rintaro commented on lib/AST/ASTDumper.cpp:
```diff
@@ -1039,29 +1039,41 @@ namespace {
       for (auto &Clause : ICD->getClauses()) {
+        OS << '\n';
+        OS.indent(Indent);
+        PrintWithColorRAII(OS, StmtColor) << (Clause.Cond ? "#if:" : "#else");
```
Should add `:` after `#else`.

rintaro committed:
[ASTScope] Eliminate isRealTopLevelCodeDecl function

Now that, all TopLevelCodeDecl is real. Just don't need this.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please test source compatibility

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2357,7 +2357,21 @@ void PrintAST::visitTopLevelCodeDecl(TopLevelCodeDecl *decl) {
+
+  for (auto &Clause : ICD->getClauses()) {
+    if (&Clause == &*ICD->getClauses().begin())
+      Printer << tok::pound_if << " "; // FIXME: print condition
```
Can you print *something* here, even a comment, so that people will know it's not a bug? Or, well, that it's a *known* bug?

rintaro replied:
I think this is just a *unimplemented* feature.
For now, added a commit for printing `/* condition */` here.

jrose-apple reviewed (commented):
Can you add a test that shows that we still reject statements within `#if` within types, and at the top level in -parse-as-library code?

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Compatibility test result
https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite/56/

CodaFi commented:
@rintaro Is there anything blocking you from merging this now?

rintaro commented:
@CodaFi 
Although I'm OK to merge this prior to #7955, I prefer to land #7955 first.
Also, this PR haven't been explicitly approved by anyone yet.

jrose-apple commented:
I didn't see any problems but I wasn't looking too closely. Slava may also still have opinions.

rintaro committed:
[ASTPrinter] Tweak #if print

* Removed extra newlines.
* Print '/* condition */' for condition part.

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[Parse] Cleanup #if directive parsing

Now that, IfConfigDecl holds ASTNode regardless statements position or
declarations postion. We can construct it in single method.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@slavapestov Could you review this?

CodaFi commented:
Just to get the validation tests in

@swift-ci please test

rintaro commented:
@slavapestov Ping
Any change to merge this?

rintaro commented:
Thanks!
